pkgs/servers/mastodon/default.nix

```diff
@@ -0,0 +1,77 @@
+{ nodejs-slim, yarn2nix-moretea, fetchFromGitHub, bundlerEnv,
+  stdenv, yarn, lib, callPackage, ... }:
+
+let
+  version = import ./version.nix;
+  src = callPackage ./source-patched.nix {};
+
+  mastodon-gems = bundlerEnv {
+    name = "mastodon-gems";
+    inherit version;
+    gemdir = src;
+    gemset = ./gemset.nix;
+  };
+
+  mastodon-js-modules = yarn2nix-moretea.mkYarnPackage {
+    name = "mastodon-modules";
+    yarnNix = ./yarn.nix;
+    packageJSON = ./package.json;
+    inherit src;
+    inherit version;
+  };
+
+  mastodon-assets = stdenv.mkDerivation {
+    name = "mastodon-assets";
+    inherit src version;
+
+    buildInputs = [
+      mastodon-gems nodejs-slim yarn
+    ];
+
+    buildPhase = ''
+      cp -r "${mastodon-js-modules}/libexec/mastodon/node_modules" "node_modules"
+      chmod -R u+w node_modules
+      rake assets:precompile
+    '';
+
+    installPhase = ''
+      mkdir -p $out/public
+      cp -r public/assets $out/public
+      cp -r public/packs $out/public
+    '';
+  };
+
+in stdenv.mkDerivation {
+  name = "mastodon";
+  inherit src version;
+
+  passthru.updateScript = callPackage ./update.nix {};
+
+  buildPhase = ''
+    ln -s ${mastodon-js-modules}/libexec/mastodon/node_modules node_modules
+    ln -s ${mastodon-assets}/public/assets public/assets
+    ln -s ${mastodon-assets}/public/packs public/packs
+
+    for b in $(ls ${mastodon-gems}/bin/)
+    do
+      rm -f bin/$b
+      ln -s ${mastodon-gems}/bin/$b bin/$b
+    done
+
```

Comment:
> It seems unlikely that overwriting them with executables from the bundled environment is going to work correctly.

Can we confirm this before complicating the code? Fwiw, we can always use patchShebangs rather than doing send.

Reply:
Just doing patchShebangs does not work for all commands.

Reply:
As a specific example, if you overwrite `bin/rails` with the `rails` from the bundled environment, then the command `rails db:migrate` which is suggested in [the Mastodon documentation](https://docs.joinmastodon.org/administration/troubleshooting/) does not work.  It does work if you use Mastodon's `bin/rails` script with the path to Ruby patched.  My previous suggestion did not patch all the instances of `#!/usr/bin/env ruby` in the `bin` directory, so here is an improved version:

```
    sed -i 's#/usr/bin/env ruby#${mastodon-gems.wrappedRuby}/bin/ruby#g' bin/*

    for b in $(ls ${mastodon-gems}/bin/)
    do
      if [ ! -f bin/$b ]; then
        ln -s ${mastodon-gems}/bin/$b bin/$b
      fi
    done
```

Reply:
What happens with `patchShebangs`?

Reply:
> What happens with `patchShebangs`?

The ruby shebang gets replaced and the script can be executed, but it fails to find any dependencies from mastodon-gems.

Reply:
> The executables in Mastodon's `bin` directory are scripts with `/usr/bin/env ruby` shebangs which load configuration from elsewhere in the source tree before doing whatever it is they do. It seems unlikely that overwriting them with executables from the bundled environment is going to work correctly. I suggest:
> 
> ```
>     for b in $(ls ${mastodon-gems}/bin/)
>     do
>       if [ -f bin/$b ]; then
>         sed -i 's#/usr/bin/env ruby#${mastodon-gems.wrappedRuby}/bin/ruby#g' bin/$b
>       else
>         ln -s ${mastodon-gems}/bin/$b bin/$b
>       fi
>     done
> ```

Even with this, tootctl fails with
```
Could not find rake-12.3.3 in any of the sources
Run `bundle install` to install missing gems.
```
To make tootctl work I had to combine the first few lines of one of the ruby scripts in mastodon-gems/bin/ with the tootctl script:

```
#!/nix/store/jcamgkl7k805gzvw8v5h9fsxi2nidqnv-ruby-2.6.5/bin/ruby

ENV["BUNDLE_GEMFILE"] = "/nix/store/fwpr79frmm1iiz4nr057fxqgi3j82kns-gemfile-and-lockfile/Gemfile"
ENV["BUNDLE_PATH"] = "/nix/store/bg5mkks6rq2imp792yyz3zbkgiz6cwyf-mastodon-gems/lib/ruby/gems/2.6.0"
ENV['BUNDLE_FROZEN'] = '1'

Gem.use_paths("/nix/store/dpwih2mskka72049rypw539xg6ih2zqn-bundler-1.17.3/lib/ruby/gems/2.6.0", ENV["GEM_PATH"])

APP_PATH = File.expand_path('/nix/store/8x81l89rc9wz6i6qbj0pw4352safwwsn-mastodon-3.0.1/config/application', __dir__)
require_relative '/nix/store/8x81l89rc9wz6i6qbj0pw4352safwwsn-mastodon-3.0.1/config/boot'
require_relative '/nix/store/8x81l89rc9wz6i6qbj0pw4352safwwsn-mastodon-3.0.1/lib/cli'
Mastodon::CLI.start(ARGV)
```

Reply:
I have to correct myself there: The wrappedRuby thing works, but in your code tootctl is not touched since it does not appear in mastodon-gems. I have committed a version that should fix all commands.